astropy/stats/sigma_clipping.py

```diff
@@ -101,281 +192,428 @@ class SigmaClip:
 
-    def _perform_clip(self, _filtered_data, axis=None):
-        """
-        Perform sigma clip by comparing the data to the minimum and
-        maximum values (median + sig * standard deviation). Use
-        sigma_lower and sigma_upper to get the correct limits. Data
-        values less or greater than the minimum / maximum values
-        will have True set in the mask array.
-        """
-
-        if _filtered_data.size == 0:
-            return _filtered_data
-
-        max_value = self.cenfunc(_filtered_data, axis=axis)
-        std = self.stdfunc(_filtered_data, axis=axis)
-        min_value = max_value - std * self.sigma_lower
-        max_value += std * self.sigma_upper
-
-        # Ensure min/max can be broadcast with the data (if arrays):
-        if axis is not None:
-            if not isiterable(axis):
-                axis = (axis,)
-
-            # Convert negative indices & restore reduced axes, with length 1:
-            axis = tuple(_filtered_data.ndim + n if n < 0 else n for n in axis)
-            mshape = tuple(1 if dim in axis else sz
-                           for dim, sz in enumerate(_filtered_data.shape))
+    def _parse_cenfunc(self, cenfunc):
+        if isinstance(cenfunc, str):
+            if cenfunc == 'median':
+                if HAS_BOTTLENECK:
+                    cenfunc = _nanmedian
+                else:
+                    cenfunc = np.nanmedian  # pragma: no cover
+
+            elif cenfunc == 'mean':
+                if HAS_BOTTLENECK:
+                    cenfunc = _nanmean
+                else:
+                    cenfunc = np.nanmean  # pragma: no cover
+
+            else:
+                raise ValueError('{} is an invalid cenfunc.'.format(cenfunc))
+
+        return cenfunc
+
+    def _parse_stdfunc(self, stdfunc):
+        if isinstance(stdfunc, str):
+            if stdfunc != 'std':
+                raise ValueError('{} is an invalid stdfunc.'.format(stdfunc))
+
+            if HAS_BOTTLENECK:
+                stdfunc = _nanstd
+            else:
+                stdfunc = np.nanstd  # pragma: no cover
+
+        return stdfunc
+
+    def _compute_bounds(self, data, axis=None):
+        # ignore RuntimeWarning if the array (or along an axis) has only
+        # NaNs
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", category=RuntimeWarning)
+            self._max_value = self.cenfunc(data, axis=axis)
+            std = self.stdfunc(data, axis=axis)
+            self._min_value = self._max_value - (std * self.sigma_lower)
+            self._max_value += std * self.sigma_upper
+
+    def _sigmaclip_noaxis(self, data, masked=True, copy=True):
+        filtered_data = data.ravel()
+
+        # remove masked values and convert to ndarray
+        if isinstance(filtered_data, np.ma.MaskedArray):
+            filtered_data = filtered_data.data[~filtered_data.mask]
+
+        # remove invalid values
+        good_mask = np.isfinite(filtered_data)
```

Comment:
In some cases, inf/-inf might be considered 'valid'. It's not clear to me if it makes more sense to have ``np.isnan`` or ``np.isfinite`` here. I hesitate to suggest adding another option - what do you think?

Reply:
If the inf/-inf is not removed/masked, then nothing will be clipped (the center will be inf and the std is nan).  Actually, in that case the code might even crash.  It was already changed from `np.isnan` to `~np.isfinite` in https://github.com/astropy/astropy/commit/149d58e6cbe3c46f3def78be46d1742a8d7566fc#diff-9a14850376ba3cb948038a9ac71a2bf6 (PR discussion:  https://github.com/astropy/astropy/pull/4051)